machinelearning/tensorflow.md

```diff
@@ -0,0 +1,415 @@
+export DNNL_DEFAULT_FPMATH_MODE=BF16
+
+# Make sure the openmp threads are distributed across all the processes for multi process applications to avoid over subscription for the vcpus.
+export OMP_NUM_THREADS = <no. of vcpus> / <no. of processes>
+export OMP_PROC_BIND=false
+export OMP_PLACES=cores
+
+```
+
+```
+# Tensorflow's inter and intra_op_parallelism_thread settings are critical for the optimal workload parallelization in a multi-threaded system.
+# set the inter and intra op thread count during the session creation
+session = Session(
+                 config=ConfigProto(
+                      intra_op_parallelism_threads=<num. of vcpus>,
+                      inter_op_parallelism_threads=1,
+                )
+)
+
+```
+
+Tensorflow recommends the graph optimization pass for inference to remove training specific nodes, fold batchnorms and fuse operators. This is a generic optimizaion across CPU, GPU or TPU inference, and the optimization script is part of the Tensorflow python [tools](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/optimize_for_inference_lib.py). For a detailed description, please refer to the Tensorflow [Grappler](https://www.tensorflow.org/guide/graph_optimization) documentation. Below is a snippet of what libraries to import and how to invoke the Grappler passes for inference.
+
+```
+from tensorflow.python.tools.optimize_for_inference_lib import optimize_for_inference
+
+graph_def = tf.compat.v1.GraphDef()
+with tf.compat.v1.gfile.FastGFile(model_path, "rb") as f:
+     graph_def.ParseFromString(f.read())
+
+optimized_graph_def = optimize_for_inference(graph_def, [item.split(':')[0] for item in inputs],
+                    [item.split(':')[0] for item in outputs], dtypes.float32.as_datatype_enum, False)
+g = tf.compat.v1.import_graph_def(optimized_graph_def, name='')
+```
+Note: While the Grappler optimizer covers majority of the networks, there are few scenarios where either the Grappler optimizer can't optimize the generic graph or the runtime kernel launch overhead is simply not acceptable. XLA addresses these gaps by providing an alternative mode of running models: it compiles the TensorFlow graph into a sequence of computation kernels generated specifically for the given model. Please refer to the **Enable XLA optimizations** section below to compile Tensorflow with the downstream optimizations and using them for inference.
+
+# Evaluate performance with standard MLPerf inference benchmarks
+
+1. Setup MLPerf inference benchmarks and the required tools.
+
+```
+sudo apt install -y build-essential cmake libgl1-mesa-glx libglib2.0-0 libsm6 libxrender1 libxext6 python3-pip
+
+git clone https://github.com/mlcommons/inference.git --recursive
+cd inference
+git checkout 5ec3ac922556107ce0f6ca63c175d379017ba3d8
+cd loadgen
+CFLAGS="-std=c++14" python3 setup.py bdist_wheel
+pip install <dist/*.whl>
```

Comment:
do you just mean `dist/*.whl` ?

Reply:
yes. but when I do that the markdown isn't looking correct. Will check how else I can fix it.